3594: Update cookie information page

holt83 opened the pull request:
#### Link to issue

https://platform.dandigbib.org/issues/3594

This work is closely linked to https://github.com/ding2/ding2/pull/1538 and should be merged together in the same release.

#### Description

Updates the cookie information page to be more compliant with [current laws](https://www.datatilsynet.dk/media/7958/vejledning-om-behandling-af-personoplysninger-om-hjemmesidebesoegende.pdf). We describe more in detail which categories of cookies we use and why we use them. Furthermore we also describe exactly which cookies are blocked, when using the blocking functionality from eu_cookie_compliance module that we're implementing in #1538. 

To avoid overriding existing information on the current cookie page, we take a backup and append the title with " (BACKUP)", before creating a new cookie page.
  
Additionally we ensure that our `ding2_set_cookie_page()` is compatible with the new paragraphs field.

#### Screenshot of the result

![3594-new-cookie-page](https://user-images.githubusercontent.com/5011234/80016214-198f2980-84d3-11ea-87d2-01997348b44d.png)

#### Checklist

- [ ] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [ ] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [ ] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

holt83 commented:
I updated the PR with Scrutinizer fixes:

1. I can't do anything about the the remaining issues in ding_webtrekk.opt-out.js.
2. I'm not sure I understand the "Doc comment short description must be on a single line, further text should be a separate paragraph" issue in ding2.install?

kasperg commented on ding2.install:
```diff
@@ -624,3 +624,28 @@ function ding2_update_7051() {
+        ),
+      ),
+    );
+    node_save($node);
```
This will override any modifications to the text that editors might have made locally. I do not know if there is anything we can do about this.

holt83 replied:
Maybe we could copy the old text to a new node?

holt83 replied:
Update the PR to make backup of old cookie page on update.

kasperg commented on modules/ding_webtrekk/ding_webtrekk.install:
```diff
@@ -0,0 +1,60 @@
+function _ding_webtrekk_set_cookie_compliance_text() {
+  $body = '<p><strong>Webtrekk</strong></p>
+    <p>Vi bruger Webtrekk til at føre statistik over trafikken på hjemmesiden. Al indsamlet statistik er anonym.<br>(<a class="external" href="https://www.webtrekk.com/en/legal/opt-out-webtrekk/">Webtrekk - om brug af cookies på websider og øvrige forhold omkring persondata</a>)<br>- Hvis du vil fravælge cookies fra Webtrekk kan du trykke her:
+    <p><a class="button webtrekk-opt-out" href="#">Fravælg Webtrekk cookie</a></p>
```
This is fragile. We are tying key functionality to a special class set in a WYSIWYG editor text field. In other words: If an editor deletes this by accident it is hard to get back.

holt83 replied:
Yes, this is unfortunate. Maybe a node isn't the right thing for this? I'll see if I can come up with a better solution.

holt83 replied:
Updated PR: Both webtrekk opt-out text and javascript is now added in hook_node_view().

kasperg commented on modules/ding_webtrekk/js/ding_webtrekk.opt-out.js:
```diff
@@ -0,0 +1,28 @@
+      // If Webtrekk is not setup correctly, we might not have this.
+      if (typeof webtrekkV3 === 'undefined') {
+        return;
+      }
```
I suggest moving this to the `click` handler itself. This will give the browser additional time to load webtrekk.

holt83 replied:
Done :+1:

holt83 commented on modules/ding_webtrekk/js/ding_webtrekk.opt-out.js:
```diff
@@ -0,0 +1,28 @@
+        webtrekk.sendinfo({linkId: 'event_optout'});
+        // Set the opt-out cookie (for 5 years).
+        webtrekk.setCookie ('webtrekkOptOut', 1, 60 * 60 * 24 * 30 * 12 * 5);
+        alert(Drupal.t('Dear user, you have successfully excluded yourself from tracking.'));
```
Yes, it's from webtrekk. Maybe a better approach would be to use ding_popup and change text to just "you have successfully excluded yourself from Webtrekk tracking".

holt83 replied:
I updated the pr to use ding_popup for confirmation.

I'm not sure about the appropiate opt-out time.

holt83 commented:
Updated the PR:

- Use ding_popup for confirmation text.
- Webtrekk opt-out text and javascript is now added in hook_node_view(), to prevent editing out html-class.
- Make backup of old cookie page on update.

Still need clarification on the appropiate opt-out time and whether or not it's ok to track by default/or track at all.

holt83 commented:
PR is rebased and updated to work with Paragraphs.

holt83 commented:
I'm getting this on circle: 

```
E: Failed to fetch http://deb.debian.org/debian/pool/main/libp/libpng1.6/libpng-tools_1.6.28-1_amd64.deb  404  Not Found
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

Will try rerun tommorow

JrgenGNielsen commented:
reviewed and approved
